### ISSUE TITLE
Marketing Book Theme - Page numbers interchange issue in Index

### ISSUE THREAD

**Issue by srm-openstax** [openstax/cnx-recipes#4855]
We have received the below feedback from customer. Please find the below screenshot. Kindly check and resolve this issue.

**In page proof, page 42 should come before p. 165 in a list of pages.**

![index-issue](https://user-images.githubusercontent.com/86622862/190337366-3e8273e6-b22f-468a-a759-7d6ba1c14870.png)

# Location(s)
- **osbooks repo:** https://github.com/openstax/osbooks-principles-marketing
- **collection slug:** principles-marketing
- **ref/branch:** https://openstax-osbooksprincip-zkvocjjsaof.ws-us65.gitpod.io/
- **chapter/section:** Index (Global issue)
- **surrounding text:** Starbucks 16, 166, 42

**Comment by wilkus27** [comment]
@kswanson33 @LostMojo I'm interested in your opinion :wink: 

Problem comes from this part of code:

![image](https://user-images.githubusercontent.com/40228854/193576625-903c928a-2995-4686-946d-257de5742402.png)

First we are adding terms from pages, then from composite pages. That's why numbering is messy. 
I tried to use this trick to iterate thru `pages` and `composite_pages` at once:

![image](https://user-images.githubusercontent.com/40228854/193577504-c72af267-26c0-4f09-98d5-4bb61ff8b666.png)

But some books have terms also in Appendix.

So what we need is to iterate thru `pages` from `book` and `composite_pages` from `book.chapters`, because we are using chapter number here:

![image](https://user-images.githubusercontent.com/40228854/193578324-488507f7-6492-49d4-868e-504a2f80d01d.png)

Should we try to modify `search_with` method somehow, so we can add parent enumerator to enumerators that we are searching with?

![image](https://user-images.githubusercontent.com/40228854/193580062-66e10a88-8575-403e-aabb-6230fc62bd79.png)

So it would be something like that:

![image](https://user-images.githubusercontent.com/40228854/193580768-a447a39d-1608-4de8-964a-b9f8dbcfa613.png)

Or is it better to leave method as is, but add terms twice: from chapters and from appendix? Appendix is at the end of book, so pages numbering won't be messy. But ids will change:

![image](https://user-images.githubusercontent.com/40228854/193581619-0b1070de-4733-4a59-bfc1-dd3255630e55.png)

Because terms will be numbering again from 1. 

Also if any book will have another EOB page with terms we have to add additional loop for that.

Which direction is better in your opinion? :upside_down_face:

**Comment by kswanson33** [comment]
@wilkus27 

> First we are adding terms from pages, then from composite pages. That's why numbering is messy.

One thought I had is what if we sorted the terms in order of the page they're on after we add them to the index? But I think this would be much more complex than just adding them in the right order to begin with.

> So it would be something like that:

```ruby
book.search_with(
  [Kitchen::PageElementEnumerator, self],
  [Kitchen::CompositePageElementEnumerator, Kitchen::ChapterElementEnumerator]
)
```

☝️ This is an interesting idea. Instead of passing in an array of single element enumerators to match, you propose to pass in an array of element trees, such that returned elements can have access to different ancestor trees. This may be possible, but it is more complex than existing traversals. If there are several places in which something like this might be the best solution, it seems like it's an option worth pursuing. But if this is the only application, I think we can come up with a workaround.

The only `pages` that exist outside of a `chapter` are `preface` and `appendix`, I believe. There are other `eob` sections, but it doesn't seem like we add terms from these anyway. So I like your idea to use different loops: maybe a `book.pages('$.preface')` loop, a `book.chapters.search_with(<page & composite page>)` loop, and a `book.pages('$.appendix')` loop. That should cover the full book content in order.

> But ids will change because terms will be numbering again from 1.

Fortunately, ids in terms are being phased out (note the `term_element.id ||= '...'` rather than `term_element.id = '...'`). I believe only Fisyka(?) actually has term IDs added by this line of code? And we're hoping to change that soon so that we can take that line out entirely. Since we're expecting to see that change shortly, and because the impact is low, don't worry about making changes to that ID. Thanks for checking in about IDs!

tl;dr I suggest to either move forward with your proposal for using element trees with `search_with`, as long as there are other instances where it'll be useful, or use the workaround above.

**Comment by wilkus27** [comment]
@kswanson33 Thank you! :relaxed: I believe this is the only place where we need such behaviour. So I'll go with workaround for now.

**Comment by MartaSztuka** [comment]
#### PASSED
##### MARKETING
![Screenshot from 2022-10-05 10-17-00.png](https://images.zenhubusercontent.com/5cb5ec14fec3b63424e0f03c/91c17748-866c-45ba-a11d-5ad35a45d2af)

##### POLITICAL SCIENCE
![Screenshot from 2022-10-05 10-40-33.png](https://images.zenhubusercontent.com/5cb5ec14fec3b63424e0f03c/8e2ccc9c-5559-4637-bad9-f142694f04bc)

##### MICROECONOMICS
![Screenshot from 2022-10-05 10-51-55.png](https://images.zenhubusercontent.com/5cb5ec14fec3b63424e0f03c/047276df-7213-496a-88ff-1d2e8089ddb5)

##### FINANCIAL ACCOUNTING
![Screenshot from 2022-10-05 11-30-14.png](https://images.zenhubusercontent.com/5cb5ec14fec3b63424e0f03c/259caf98-049a-4fa1-9e33-919c93bd9d98)
